api/prometheus.go

```diff
@@ -0,0 +1,453 @@
+			return 0, fmt.Errorf("cannot parse %q to a valid duration. It overflows int64", s)
+		}
+		return time.Duration(ts), nil
+	}
+	if d, err := model.ParseDuration(s); err == nil {
+		return time.Duration(d), nil
+	}
+	return 0, fmt.Errorf("cannot parse %q to a valid duration", s)
+}
+
+// Select returns a set of series that matches the given label matchers.
+func (q *querier) Select(matchers ...*labels.Matcher) (storage.SeriesSet, error) {
+	minFrom := uint32(math.MaxUint32)
+	var maxTo uint32
+	var target string
+	var reqs []models.Req
+
+	expressions := []string{}
+	for _, matcher := range matchers {
+		if matcher.Name == model.MetricNameLabel {
+			matcher.Name = "name"
+		}
+		if matcher.Type == labels.MatchNotRegexp {
+			expressions = append(expressions, fmt.Sprintf("%s!=~%s", matcher.Name, matcher.Value))
+		} else {
+			expressions = append(expressions, fmt.Sprintf("%s%s%s", matcher.Name, matcher.Type, matcher.Value))
+		}
+	}
+
+	series, err := q.clusterFindByTag(q.ctx, q.OrgID, expressions, 0)
+	if err != nil {
+		return nil, err
+	}
+
+	if q.metadataOnly {
+		return BuildMetadataSeriesSet(series)
+	}
+
+	minFrom = util.Min(minFrom, q.from)
+	maxTo = util.Max(maxTo, q.to)
+	for _, s := range series {
+		for _, metric := range s.Series {
+			for _, archive := range metric.Defs {
+				consReq := consolidation.None
+				fn := mdata.Aggregations.Get(archive.AggId).AggregationMethod[0]
+				cons := consolidation.Consolidator(fn)
+
+				newReq := models.NewReq(archive.Id, archive.NameWithTags(), target, q.from, q.to, math.MaxUint32, uint32(archive.Interval), cons, consReq, s.Node, archive.SchemaId, archive.AggId)
+				reqs = append(reqs, newReq)
+			}
+		}
+	}
+
+	select {
+	case <-q.ctx.Done():
+		//request canceled
+		return nil, fmt.Errorf("request canceled")
+	default:
+	}
+
+	reqRenderSeriesCount.Value(len(reqs))
+	if len(reqs) == 0 {
+		return nil, fmt.Errorf("no series found")
+	}
+
+	// note: if 1 series has a movingAvg that requires a long time range extension, it may push other reqs into another archive. can be optimized later
+	reqs, _, _, err = alignRequests(uint32(time.Now().Unix()), minFrom, maxTo, reqs)
+	if err != nil {
+		log.Error(3, "HTTP Render alignReq error: %s", err)
+		return nil, err
+	}
+
+	out, err := q.getTargets(q.ctx, reqs)
+	if err != nil {
+		log.Error(3, "HTTP Render %s", err.Error())
+		return nil, err
+	}
+
+	return SeriesToSeriesSet(out)
+}
+
+// LabelValues returns all potential values for a label name.
+func (q *querier) LabelValues(name string) ([]string, error) {
+	expressions := []string{"name=~[a-zA-Z_][a-zA-Z0-9_]*$"}
+	if name == model.MetricNameLabel {
+		name = "name"
+		expressions = append(expressions, "name=~[a-zA-Z_:][a-zA-Z0-9_:]*$")
+	}
+	result, err := q.MetricIndex.FindTagValues(q.OrgID, name, "", expressions, 0, 100000)
+	if err != nil {
+		return nil, fmt.Errorf("unable to fetch label values for key:%s, error: %v", name, err)
+	}
+	if result == nil {
+		result = []string{}
+	}
```

Comment:
i bet you don't need these 3 lines. `[]string{}` is equivalent to a `nil` string slice.